smartconfig/configuration.py

```diff
@@ -0,0 +1,173 @@
+import inspect
+from itertools import chain
+from typing import Any, Dict, List, Optional, Tuple
+
+from smartconfig.exceptions import ConfigurationKeyError, DuplicateConfiguration, InvalidOperation
+from smartconfig.parser import YamlLikeParser
+from smartconfig.register import _register
+from smartconfig.typehints import _EntryMapping, _FilePath
+
+
+def load_config_file(path: _FilePath) -> None:
+    """
+    Load a YAML-like configuration file and update configuration entries.
+
+    The values set in the YAML-like file will override values already defined in the `ConfigEntry`.
+    For each section, the name of previous sections will be concatenated in order to make the full path of the entry
+    to override.
+
+    See `smartconfig.YamlLikeParser` for more details on the syntax.
+
+    Args:
+        path: The path to the configuration file. Can be a string or an object defining `os.PathLike`.
+
+    Raises:
+        MalformedYamlFile: The configuration file is malformed.
+        FileNotFoundError: The configuration file doesn't exist.
+        IOError: An error occurred when reading the file.
+    """
+    with open(path) as file:
+        patch = YamlLikeParser(file.read(), file.name).parse()
+
+    for path, entries in patch.items():
+        # Patch every already existing entries.
+        if path in _register.configuration_for_module:
+            _register.configuration_for_module[path]._patch_entries(entries)
+
+        # Update the global register for future entries to use.
+        if path not in _register.global_configuration:
+            _register.global_configuration[path] = {}
+        _register.global_configuration[path].update(entries)
+
+
+class _ConfigEntryMeta(type):
+    """
+    Metaclass used to define special ConfigEntry behaviors.
+
+    Note: Using this metaclass outside of the library is currently not supported.
+    """
+
+    def __getattribute__(cls, item: str):
+        """
+        Special attribute lookup function.
+
+        If the attribute name starts with `_`, normal lookup is done, otherwise _configuration_mapping is used.
+
+        Args:
+            item: Attribute to lookup
+
+        Raises:
+            ConfigurationKeyError: The item doesn't exist.
+        """
+        # Use the normal lookup for attribute starting with `_`
+        if item.startswith('_'):
+            return super().__getattribute__(item)
+
+        if item not in cls._configuration_mapping:
+            raise ConfigurationKeyError(f"Entry {cls._path} doesn't define any {item} entry.")
+
+        return cls._configuration_mapping[item]
+
+    def __new__(cls, name: str, bases: Tuple[type, ...], dict_: Dict[str, Any]) -> type:
+        """Add special attribute to the new entry."""
+        configuration_mapping = {key: value for key, value in dict_.items() if not key.startswith('_')}
```

Comment:
> Storing and overriding defaults means that the config has to be re-patched if a default value is changed dynamically or if a module containing a class is reloaded. This can be avoided by attempting to look up the value in the config file's contents and falling back to super().__getattribute__ if not found. This would entirely eliminate the need to patch since the user-configured values could be looked up in the global dictionary.

This is a very good point, and it has been changed in 5f8c07bc04696475375293782c7e47bab57db6a0.

> Anyway, I don't think the fundamental approach of how you store things is going to work out if something like `_SUFFIX_NODES` is to be supported. This is because it breaks the assumption that all attributes of a class are going to share the same node in YAML too.
> 
> This is something that needs to be discussed further. My approach of translating the attribute to a node path and looking up the node works but it may or may not have a significant performance disadvantage since it has to look up the node every time `getattribute` is invoked. Maybe it's possible to pre-compute these lookups and store them in a dictionary, but that would have to be kept up to date somehow every time either a config file is loaded or the class is reloaded.
> 
> Of course, we can discuss if we even want to support remapping like `_SUFFIX_NODES`.

We could directly store the output of the YAML parser and use a default implementation that would walk through the tree, but like you said this has a runtime cost. While we could catch that using something like the `_configuration_mapping` we had previously, it might cause some issues if the lookup is done based on a global variable or so, as changes to this variable won't invalidate the cache. We could also provide a function to recompute the cache, but honestly, I'm not sure if it is worth it. In which cases do you think we could benefit from using a custom lookup function?

Reply:
> In which cases do you think we could benefit from using a custom lookup function?

The idea was to use it to aggregate all Discord objects of the same type in the YAML config file keeping them local to their respective modules in Python. This would make it easier to manage them in the config under the presumption that they would most often be managed in tandem rather than individually.

Furthermore, keep in mind that the bootstrapper will need some way to access all the Discord objects. It will essentially need to aggregate them anyway, the difference being that it needs all related attribute names - not just the user-configured ones.

> walk through the tree, but like you said this has a runtime cost

Yes but we need to not get ahead of ourselves. How bad would these costs really be?

> We could also provide a function to recompute the cache, but honestly, I'm not sure if it is worth it.

Believe me, I spent a lot of time thinking about it before writing that comment. Here's how I imagined an implementation:

After a config is loaded, each config class is iterated. A map is made with the attribute name as the key and the value coming from the YAML. If there's nothing in YAML for that attribute, then the key isn't stored in the map. The way it gets the value from the YAML is by calling a function on the config class which returns the corresponding YAML path for a given attribute name. That will be used to walk through the nodes to retrieve the value in the config similarly to my original implementation only it's stored in a map rather than looked up dynamically.

The entire cache would have to be re-built every time the config file is reloaded. It'd also have to be rebuilt when a class is reloaded i.e. re-imported (let's not worry about dynamically adding/removing attributes from the class). Each class could have its own cache so that the rest of the cache won't need to also be re-generated when a single class changes.

This is indeed similar to what you were doing but the differences are that it doesn't store default values and that the YAML value can be looked for everywhere - not just directly under the node for that module and with the exact same attribute name.

> it might cause some issues if the lookup is done based on a global variable or so, as changes to this variable won't invalidate the cache.

I'm not sure what you mean. The value from the config would be stored directly in the cache local to the class.

Reply:
Alright, I think I see what you want with the custom lookup path for Discord IDs. That said, instead of using a custom lookup path, why not do something like that?

```py
class DiscordChannelEntry:
    def __init__(self, channel_name: str):
        self._channel_name = channel_name

    def __get__(self):
        return registry.global_configuration["discord.channels"][self._channel_name]
        # Maybe use a public function?

    def __set__(self, *_):
        raise ...

### duckpond.py ###

class Configuration(ConfigEntry):
    required_duckies: int = 4
    duckpond_channel: int = DiscordChannelEntry('duckpond')
```

Here are the pros and cons I could see from this approach instead of using a custom lookup path:

Pros:
* It is more explicit, you can directly see that by default the Channel lookup will be done differently, you don't have to guess that there's a custom lookup path here.
* You can still override the attribute using the full dotted path, like `bot.cogs.fun.duckpond.duckpond_channel`, If you want to do that for any other reason.

Cons:
* You need to add the descriptor, you can't just rely on attribute names.

Note that I'm not saying that custom lookup is something that shouldn't be implemented, but I think it is rather something we would want to support at a later time, for now, we may not need it depending on which method we choose to use. 

What do you think?

Reply:
This was a consideration but I prefer more abstraction. It'd also become redundant - it's more convenient to specify a mapping once and globally apply it to all config classes than to explicitly set it on every single attribute.